File: picar-client-android/app/src/main/java/com/rpinferetti/picar/Gamepad.java
package com.rpinferetti.picar;

import android.view.InputDevice;
import android.view.KeyEvent;
import android.view.MotionEvent;

public class Gamepad {

    private GamepadMap mGamepadMap;
    private OnGamepadListener mListener;

    public Gamepad() {
        mGamepadMap = new GamepadMap();
    }

    public void handleButtonPressed(KeyEvent event) {
        int action = event.getAction();
        int keyCode = event.getKeyCode();

        switch (keyCode) {
            case KeyEvent.KEYCODE_BUTTON_X:
                if (action == KeyEvent.ACTION_DOWN)
                    mGamepadMap.setButtonX(true);
                else
                    mGamepadMap.setButtonX(false);
                break;

            case KeyEvent.KEYCODE_BUTTON_A:
                if (action == KeyEvent.ACTION_DOWN)
                    mGamepadMap.setButtonA(true);
                else
                    mGamepadMap.setButtonA(false);
                break;

            case KeyEvent.KEYCODE_BUTTON_Y:
                if (action == KeyEvent.ACTION_DOWN)
                    mGamepadMap.setButtonY(true);
                else
                    mGamepadMap.setButtonY(false);
                break;

            case KeyEvent.KEYCODE_BUTTON_B:
                if (action == KeyEvent.ACTION_DOWN)
                    mGamepadMap.setButtonB(true);
                else
                    mGamepadMap.setButtonB(false);
                break;

            case KeyEvent.KEYCODE_BUTTON_R1:
                if (action == KeyEvent.ACTION_DOWN)
                    mGamepadMap.setButtonR1(true);
                else
                    mGamepadMap.setButtonR1(false);
                break;

            case KeyEvent.KEYCODE_BUTTON_L1:
                if (action == KeyEvent.ACTION_DOWN)
                    mGamepadMap.setButtonL1(true);
                else
                    mGamepadMap.setButtonL1(false);
                break;
        }

        mListener.onGamepadMapChanged(mGamepadMap);
    }

    private float getCenteredAxis(MotionEvent event, InputDevice device,
                                  int axis, int historyPos) {
        final InputDevice.MotionRange range = device.getMotionRange(axis, event.getSource());

        // A joystick at rest does not always report an absolute position of
        // (0,0). Use the getFlat() method to determine the range of values
        // bounding the joystick axis center.
        if (range != null) {
            final float flat = range.getFlat();
            final float value = historyPos < 0 ? event.getAxisValue(axis) :
                    event.getHistoricalAxisValue(axis, historyPos);

            // Ignore axis values that are within the 'flat' region of the
            // joystick axis center.
            if (Math.abs(value) > flat) {
                return value;
            }
        }
        return 0;
    }

    public void processJoystickInput(MotionEvent event, int historyPos) {
        InputDevice inputDevice = event.getDevice();

        // Left control stick and hat
        // Horizontal distance
        float lx = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_X, historyPos);
        if (lx == 0) {
            lx = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_HAT_X, historyPos);
        }
        // Vertical distance
        float ly = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_Y, historyPos);
        if (ly == 0) {
            ly = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_HAT_Y, historyPos);
        }

        // Right control stick
        // Horizontal distance
        float rx = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_Z, historyPos);
        // Vertical distance
        float ry = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_RZ, historyPos);

        // Shoulder triggers
        // Left
        float lt = historyPos < 0 ? event.getAxisValue(MotionEvent.AXIS_BRAKE) :
                event.getHistoricalAxisValue(MotionEvent.AXIS_BRAKE, historyPos);
        // Right
        float rt = historyPos < 0 ? event.getAxisValue(MotionEvent.AXIS_GAS) :
                event.getHistoricalAxisValue(MotionEvent.AXIS_GAS, historyPos);

        mGamepadMap.setLeftStickX(lx * 100);
        mGamepadMap.setLeftStickY(ly * 100);
        mGamepadMap.setRightStickX(rx * 100);
        mGamepadMap.setRightStickY(ry * 100);
        mGamepadMap.setLeftShoulderTrigger(lt * 100);
        mGamepadMap.setRightShoulderTrigger(rt * 100);

        mListener.onGamepadMapChanged(mGamepadMap);
    }

    public void setOnGamepadListener(OnGamepadListener listener) {
        mListener = listener;
    }

    public interface OnGamepadListener {
        void onGamepadMapChanged(GamepadMap map);
    }
}

File: app/src/main/java/com/rpinferetti/picar/Gamepad.kt
package com.rpinferetti.picar

import android.view.InputDevice
import android.view.KeyEvent
import android.view.MotionEvent
import kotlin.math.abs

class Gamepad {
    private val mGamepadMap: GamepadMap = GamepadMap()
    private var mListener: OnGamepadListener? = null

    fun handleButtonPressed(event: KeyEvent) {
        val action = event.action

        when (event.keyCode) {
            KeyEvent.KEYCODE_BUTTON_X -> mGamepadMap.isButtonX = action == KeyEvent.ACTION_DOWN

            KeyEvent.KEYCODE_BUTTON_A -> mGamepadMap.isButtonA = action == KeyEvent.ACTION_DOWN

            KeyEvent.KEYCODE_BUTTON_Y -> mGamepadMap.isButtonY = action == KeyEvent.ACTION_DOWN

            KeyEvent.KEYCODE_BUTTON_B -> mGamepadMap.isButtonB = action == KeyEvent.ACTION_DOWN

            KeyEvent.KEYCODE_BUTTON_R1 -> mGamepadMap.isButtonR1 = action == KeyEvent.ACTION_DOWN

            KeyEvent.KEYCODE_BUTTON_L1 -> mGamepadMap.isButtonL1 = action == KeyEvent.ACTION_DOWN
        }

        mListener?.onGamepadMapChanged(mGamepadMap)
    }

    private fun getCenteredAxis(
        event: MotionEvent, device: InputDevice,
        axis: Int, historyPos: Int
    ): Float {
        val range = device.getMotionRange(axis, event.source)

        // A joystick at rest does not always report an absolute position of
        // (0,0). Use the getFlat() method to determine the range of values
        // bounding the joystick axis center.
        if (range != null) {
            val flat = range.flat
            val value = if (historyPos < 0)
                event.getAxisValue(axis)
            else
                event.getHistoricalAxisValue(axis, historyPos)

            // Ignore axis values that are within the 'flat' region of the
            // joystick axis center.
            if (abs(value) > flat) {
                return value
            }
        }
        return 0f
    }

    fun processJoystickInput(event: MotionEvent, historyPos: Int) {
        val inputDevice = event.device

        // Left control stick and hat
        // Horizontal distance
        var lx = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_X, historyPos)
        if (lx == 0f) {
            lx = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_HAT_X, historyPos)
        }
        // Vertical distance
        var ly = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_Y, historyPos)
        if (ly == 0f) {
            ly = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_HAT_Y, historyPos)
        }

        // Right control stick
        // Horizontal distance
        val rx = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_Z, historyPos)
        // Vertical distance
        val ry = getCenteredAxis(event, inputDevice, MotionEvent.AXIS_RZ, historyPos)

        // Shoulder triggers
        // Left
        val lt = if (historyPos < 0)
            event.getAxisValue(MotionEvent.AXIS_BRAKE)
        else
            event.getHistoricalAxisValue(MotionEvent.AXIS_BRAKE, historyPos)
        // Right
        val rt = if (historyPos < 0)
            event.getAxisValue(MotionEvent.AXIS_GAS)
        else
            event.getHistoricalAxisValue(MotionEvent.AXIS_GAS, historyPos)

        mGamepadMap.leftStickX = lx * 100
        mGamepadMap.leftStickY = ly * 100
        mGamepadMap.rightStickX = rx * 100
        mGamepadMap.rightStickY = ry * 100
        mGamepadMap.leftShoulderTrigger = lt * 100
        mGamepadMap.rightShoulderTrigger = rt * 100

        mListener?.onGamepadMapChanged(mGamepadMap)
    }

    fun setOnGamepadListener(listener: OnGamepadListener) {
        mListener = listener
    }

    interface OnGamepadListener {
        fun onGamepadMapChanged(map: GamepadMap)
    }
}
